From 5bc489e9bf8c05e172a5ddd98ce8371b8395ed2d Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 2 May 2014 14:36:52 -0400
Subject: [misc] mei: move host_clients_map cleanup to device init

Message-id: <1399041470-21330-20-git-send-email-prarit@redhat.com>
Patchwork-id: 79307
O-Subject: [RHEL7.1 PATCH BZ 1088953 19/77] mei: move host_clients_map cleanup to device init
Bugzilla: 1088953
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1088953

commit eec86b8e85309a2cb0b33bcb361c67d81ebed474
Author: Alexander Usyskin <alexander.usyskin@intel.com>
Date:   Thu Oct 17 17:20:22 2013 +0300

    mei: move host_clients_map cleanup to device init

    Move host_clients_map cleanup from host client init to device init.

    This fixes bug where we cleaned up the bitmask that servers as pool for
    host client ids while file descriptors are kept open during suspend.
    On resume a new connection will be assigned id that is already taken
    by opened file descriptor. Consequently read/write will fail due to
    mismatched book keeping.

    Signed-off-by: Alexander Usyskin <alexander.usyskin@intel.com>
    Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Cc: Dean Nelson <dnelson@redhat.com>
Cc: Stefan Assmann <sassmann@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/misc/mei/client.c b/drivers/misc/mei/client.c
index a48c0e7..0ccc22c 100644
--- a/drivers/misc/mei/client.c
+++ b/drivers/misc/mei/client.c
@@ -365,15 +365,6 @@ void mei_host_client_init(struct work_struct *work)
 
  mutex_lock(&dev->device_lock);
 
- bitmap_zero(dev->host_clients_map, MEI_CLIENTS_MAX);
- dev->open_handle_count = 0;
-
- /*
-  * Reserving the first three client IDs
-  * 0: Reserved for MEI Bus Message communications
-  */
- bitmap_set(dev->host_clients_map, 0, 1);
-
  for (i = 0; i < dev->me_clients_num; i++) {
   client_props = &dev->me_clients[i].props;
 
diff --git a/drivers/misc/mei/init.c b/drivers/misc/mei/init.c
index 07e7ba2..64fd949 100644
--- a/drivers/misc/mei/init.c
+++ b/drivers/misc/mei/init.c
@@ -68,6 +68,14 @@ void mei_device_init(struct mei_device *dev)
  mei_io_list_init(&dev->amthif_cmd_list);
  mei_io_list_init(&dev->amthif_rd_complete_list);
 
+ bitmap_zero(dev->host_clients_map, MEI_CLIENTS_MAX);
+ dev->open_handle_count = 0;
+
+ /*
+  * Reserving the first client ID
+  * 0: Reserved for MEI Bus Message communications
+  */
+ bitmap_set(dev->host_clients_map, 0, 1);
 }
 EXPORT_SYMBOL_GPL(mei_device_init);
 
-- 
1.7.1